### PR TITLE
[BuildSystem] Handle relative paths in 'shell' makefile-style dependencies

### PR DESCRIPTION
**Pull request by dmbryson**
Under some situations clang will emit relative paths for included
dependencies. This changes 'shell' commands that use makefile-style
deps files to attempt to make dependecy paths absolute before generating
the node key.

rdar://problem/44463222

### CODE REVIEW

**Comment by dmbryson** [comment]
@swift-ci please smoke test

File: lib/BuildSystem/BuildSystem.cpp [hunk 1]
```diff
@@ -2074,7 +2075,24 @@ class ShellCommand : public ExternalCommand {
...
+          bsci.taskDiscoveredDependency(task, BuildKey::makeNode(unescapedWord));
+          return;
+        }
+
+        // Generate absolute path
+        //
+        // NOTE: This is making the assumption that relative paths coming in a
+        // dependency file are in relation to the explictly set working
+        // directory, or the current working directory when it has not been set.
+        SmallString<PATH_MAX> absPath = StringRef(command->workingDirectory);
```
Comment (dmbryson) [inline lib/BuildSystem/BuildSystem.cpp]: Hmm... this isn't quite absolute just yet.  It is assuming that 'workingDirectory' is absolute, which may not be the case.

Comment (ddunbar) [reply lib/BuildSystem/BuildSystem.cpp]: Are we sure this isn't a clang bug? If clang generates relative paths, and something uses -fworking-directory, could we ever handle that case correctly?

Comment (ddunbar) [reply lib/BuildSystem/BuildSystem.cpp]: In any case, I agree we should handle this case, just wondering if there is a corner case we can't fix

**Comment by dmbryson** [comment]
@swift-ci please smoke test

File: lib/BuildSystem/BuildSystem.cpp (hunk 1 above)
Comment (ddunbar) [reply lib/BuildSystem/BuildSystem.cpp]: Should we just make sure that every command's working directory is absolute?

File: lib/BuildSystem/BuildSystem.cpp [hunk 2]
```diff
@@ -2074,7 +2075,24 @@ class ShellCommand : public ExternalCommand {
...
+        if (!absPath.empty()) {
+          absPath.append("/");
+        }
+        absPath.append(unescapedWord);
+        llvm::sys::fs::make_absolute(absPath);
```
Comment (ddunbar) [inline lib/BuildSystem/BuildSystem.cpp]: Doesn't llvm::sys::fs have support for appending a path, versus the manual code above?

File: tests/BuildSystem/Build/discovered-makefile-deps-relative.llbuild [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# Check a build that modifies the header.
+#
+# RUN: echo "mod" >> %t.build/header-1
+# RUN: %{llbuild} buildsystem build --serial --chdir %t.build &> %t2.out
+# RUN: %{FileCheck} --check-prefix=CHECK-AFTER-MOD --input-file=%t2.out %s
+#
+# CHECK-AFTER-MOD: CC output-1
+# CHECK-AFTER-MOD: cat output-1 > ../output
+#
+# REQUIRES: platform=Darwin
```
Comment (ddunbar) [inline tests/BuildSystem/Build/discovered-makefile-deps-relative.llbuild]: Why does this require Darwin?

Comment (dmbryson) [reply tests/BuildSystem/Build/discovered-makefile-deps-relative.llbuild]: Its using the working-directory support, which is not yet implemented on linux (requires #400 to be finished).

File: lib/BuildSystem/BuildSystem.cpp (hunk 1 above)
Comment (dmbryson) [reply lib/BuildSystem/BuildSystem.cpp]: Probably not a bad idea.

File: lib/BuildSystem/BuildSystem.cpp (hunk 2 above)
Comment (dmbryson) [reply lib/BuildSystem/BuildSystem.cpp]: It does indeed.  I looked for it before, but I was looking for "join" rather than "append" and overlooked it.

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Comment by dmbryson** [comment]
@swift-ci please smoke test

File: lib/BuildSystem/BuildSystem.cpp (hunk 1 above)
Comment (jakepetroules) [reply lib/BuildSystem/BuildSystem.cpp]: Agreed, I don't think it makes any sense to allow relative working directories given that the manifest is generally intended to be entirely non-relocatable.

File: tests/BuildSystem/Build/discovered-makefile-deps-relative.llbuild (hunk 3 above)
Comment (jakepetroules) [reply tests/BuildSystem/Build/discovered-makefile-deps-relative.llbuild]: Ah yes, I should deal with that...
